[Kyu] 미션3: 체스판 초기화 / 완료

kyupid opened the pull request:
안녕하세요. 월요일 호눅스 수업에서 배운대로 커밋을 최대한 자세하게 작성했습니다.

가장 피드백 받고 싶은 부분 3개가 있습니다.

1. main() 메서드에서 스캐너로 명령어를 입력하는데 개선이 필요할까요? 스스로 그렇게 느낀 이유는 처음에 start 를 입력해서 게임을 시작하거나 end 를 입력해서 게임을 종료하도록 했는데, start 입력하고 들어가면 `if(end)` 에 접근이 안되니 start 블락 안에 또 `if(end)`를 만들었습니다. 똑같은 코드 뭉치 두개가 들어오니까 뭔가 틀린 느낌이 나서 개선이 필요하다고 생각했는데 다른 방법은 생각이 나지않습니다. -> 커밋 `Add a method print() to print` [54a7f24](https://github.com/codesquad-members-2021/java-chess/pull/86/commits/54a7f24cca8ce3cca73318a95202d5f4fdf86aa7)

2. 커밋 `Add a method to initialize a list of the white and black pawns` [310cef1](https://github.com/codesquad-members-2021/java-chess/pull/86/commits/310cef1dab0372d340ad1817d5cdb918089d6a60)

3. 커밋 `Add methods to put pawns together on one line` [a72737d](https://github.com/codesquad-members-2021/java-chess/pull/86/commits/a72737d33ba608e18f65523da7ed2ef2d917db27)

kyupid committed:
Fix test code to verify new constants before printing them out

- 테스트 코드를 먼저 작성해서 쉽게 프로덕트 코드를 생성하도록 했습니다.

- Representation 의 의미는 화면에 출력할 때 쓰이는 문자를 특정/대표한다는 의미에서 그렇게 변수를 정의하였습니다.

- 일치하는 것을 확인하기 위해 AssertJ의 assertThat() 메소드를 사용하였습니다. 일관성을 주기위해 JUnit 의 Assert 는 사용하지 않았습니다.

kyupid committed:
Fix the name of the test method to clarify its meaning better

- 메소드의 이름을 최대한 길게 서술하여 테스트 사용자가 이해하기 편하도록 수정하였습니다.

- `@DisplayName` 을 삭제하였습니다. 메소드 이름이 자세하기 때문에 굳이 `@DisplayName` 을 덧붙일 필요 없다고 생각합니다.

kyupid committed:
Add constants and a method to product code along with test code

- representation 변수와 상수가 생기면서 생성자를 수정하고 representation 을 반환하는 메소드도 추가했습니다.

- representation 변수가 생기면서 PawnTest.verifyPawn() 메소드의 파라미터를 수정했습니다.

kyupid committed:
Add a test method to verify that creating pawns with List work well

- 체스판을 실제로 생성하기 위해 초기화 작업을 하는 메소드를 테스트하기 위한 메소드를 먼저 추가합니다.

kyupid committed:
Add a method to initialize a list of the white and black pawns

- Pawn 들을 담기 위해서 ArrayList 를 선언했습니다.

- ArrayList 를 List 로 담는 게 좋다고해서 그렇게 할까 생각하다가 ArrayList로 받아야 나중에(?) List<> board에 초기화한 whitePawns 와 blackPawns 를 담을 수 있을거라고 생각했습니다.

- for문을 이용해서 white and black Pawns 을 초기화했습니다. 각 각의 체스말들은 독립적인 객체이기 때문에 모두 서로 다른 새로운 객체들로 생성해서 담았습니다.

kyupid committed:
Add methods to put pawns together on one line

- ArrayList에 담은 Pawn들을 화면에 출력하기 위해서 이전에 만든 REPRESENTATION을 이용해서 get 메소드를 포함 메소드들을 만들었습니다.

- 먼저 listPawnsToSingleLine의 파라미터 이름을 어떻게 정할지 모르겠습니다. 일단 생각나는대로 처음에 pawnsColor 라고 했다가 의미가 맞지 않는 거 같아서 whichGroupOfPawns 로 변경했습니다.

- 파라미터를 넣도록 지정해줘서 흰색이든 검은색이든 들어오면 똑같은 코드에서 문자열이 반환되도록 했습니다. 코드의 재사용성을 높였습니다.

- for문을 이용해서 ArrayList 에 들어간 각 Pawn 객체들의 REPRESENTATION을 받아서 StringBuilder에 append하고 반환했습니다.

- getPawnsResult 메소드를 각 컬러에 맞는 문자열을 반환하도록 만들었습니다

kyupid committed:
Add a constant for length of board

- 나중에 더 쓰일 것 같기도하고 가독성을 위해서 for문에 그냥 8로 넣었던 것을 상수로 만들어서 교체했습니다.

kyupid committed:
Fix a test method `verifyBoard`

- 먼저 테스트하는 메소들을 통일하기 위해서 AssertJ 의 assertThat 으로 모두 바꿨다.

- 생성자 파라미터로 color 와 representation 을 받음으로 테스트 메소드에도 representation 을 추가해서 그에 따라 pawnColor 인스턴스 생성시에도 바꾸었다.

kyupid committed:
Fix the name of the test method to clarify its meaning better.

- 가독성을 위해 메소드 이름을 구체적으로 변경했다.

kyupid committed:
Add a method print() to print

- Pawn 을 List 에 담았고, 화면에 출력하기 위해 print()를 추가했다.

- Pawn 만 추가했음으로, 나머지 빈공간을 표현하기위해 따로 공백을 뜻하는 `.`을 반환하는 메소드도 추가했다.

kyupid committed:
Add main() to start a game

- start 를 입력하면 보드가 생성 되고 초기화합니다. 그리고 while 문으로 `move` 할때마다 보드의 모습을 출력하도록 했습니다. 나중에 다른 chess pieces 들도 여기에서 표현 될 것입니다.

- end 를 입력하면 시스템을 종료합니다.

- start 했을 시 end 를 입력하면 종료하는 조건을 다시 입력해주었습니다. 이 부분은 짐작이지만 수정이 가능할 것 같은데 아직은 어떻게 해야할지 모르겠습니다.

kyupid committed:
Add a test method to verify Board.print()

- 처음에 메인 메소드에서 바로 print()를 실행했으나 에러가 떠서 곧장 테스트 메소드를 만들었습니다.

kyupid committed:
Fix access modifiers in test methods

- 테스트 메소드들의 접근 제어자를 모두 디폴트로 변경했습니다

wheejuni commented on src/main/java/net/honux/chess/Board.java:
```diff
@@ -2,13 +2,17 @@
 
     private List<Pawn> board = new ArrayList<>();
 
+    private ArrayList<Pawn> whitePawns = new ArrayList<>();
+    private ArrayList<Pawn> blackPawns = new ArrayList<>();
```
`- ArrayList 를 List 로 담는 게 좋다고해서 그렇게 할까 생각하다가 ArrayList로 받아야 나중에(?) List<> board에 초기화한 whitePawns 와 blackPawns 를 담을 수 있을거라고 생각했습니다.` 가 이해가 안 되네요. 조금 더 설명해주세요.

kyupid replied:
>ArrayList 를 List 로 담는 게 좋다고해서 그렇게 할까 생각하다가

[지난번 리뷰 때](https://github.com/codesquad-members-2021/java-chess/pull/46#discussion_r573748561) 받은 내용을 언급했습니다.

> ArrayList로 받아야 나중에(?) List<> board에 초기화한 whitePawns 와 blackPawns 를 담을 수 있을거라고 생각했습니다.

"whitePawns 와 blackPawns 를 10번 라인 `private List<Pawn> board = new ArrayList<>();`에 넣는다" 라고 생각했었는데, 지금 생각해보니 말이 맞지가 않네요. 어떻게 넣는다는 건지 저도 이해가 안갑니다. 담부터 좀 더 생각을 정리해서 작성하겠습니다.

아마도 제가 말하려는 의도는 whitePawns 와 blackPawns 를 ArrayList에 담고, 담은 Pawns 들을 또 List에 담는다는 생각으로 말했던 거 같습니다.

kyupid replied:
@wheejuni 

>아마도 제가 말하려는 의도는 whitePawns 와 blackPawns 를 ArrayList에 담고, 담은 Pawns 들을 또 List에 담는다는 생각으로 말했던 거 같습니다.

이 부분은 다른 분 PR 보고 발견했는데 [링크와 같이](https://github.com/codesquad-members-2021/java-chess/pull/75/files/ace3ace6d947813b88b9dd2a251c9aa3bf373fe8#r577073239) 표현하고 싶었던 거 같습니다.

wheejuni replied:
네, `List` 에 `List` 를 담아도 되고 이는 흔한 구현입니다. 코드가 조금 못생기긴 했죠?
주저없이 담으셔도 됩니다.

wheejuni commented on src/main/java/net/honux/chess/Board.java:
```diff
@@ -2,13 +2,17 @@
 
 public class Board {
 
     private List<Pawn> board = new ArrayList<>();
```
이 변수는 유지되어야 하나요?

kyupid replied:
가독성을 위해서 for문에 그냥 8로 넣었던 것을 상수로 만들어서 교체했습니다. 굳이 필요없는 건지 여쭤보고 싶습니다.
참고 커밋: ab4d0d5

kyupid replied:
아, 제가 초록색 부분만 보고 상수를 말씀하시는 줄 알았습니다. 12번 라인 `private List<Pawn> board = new ArrayList<>();` 말씀하시는 거였죠?

Pawn을 데이터타입으로 받는 List는 더이상 아래 whitePawns와 blackPawns와 중복되므로 필요없는 거 같습니다. 관련된 다른 것들도 수정하겠습니다

wheejuni commented on src/main/java/net/honux/chess/Board.java:
```diff
@@ -20,4 +25,55 @@ public int size() {
+    public String listPawnsToSingleLine(ArrayList<Pawn> whichGroupOfPawns) {
+        StringBuilder sb = new StringBuilder();
+
+        for (int i = 0; i < BOARD_LENGTH; i++) {
+            sb.append(whichGroupOfPawns.get(i).getRepresentation());
+        }
```
[enhanced for loop](https://jamesdreaming.tistory.com/105)를 학습하고 적용해 주세요.

kyupid replied:
감사합니다. 공부해서 적용하였습니다.

wheejuni commented on src/main/java/net/honux/chess/Board.java:
```diff
@@ -20,4 +25,55 @@ public int size() {
+        }
+    }
+
+    public String listPawnsToSingleLine(ArrayList<Pawn> whichGroupOfPawns) {
```
이 메소드는 외부에 노출될 필요가 없어 보이네요. `public` 은 과한 것 같습니다.

kyupid replied:
외부에 노출될 필요가 없다는 말씀은 패키지 밖으로 노출될 필요가 없다는 말씀이신가요 아니면 클래스 내부에서만 써야 된다는 말씀이신가요?

main 메소드에서 initialize()를 사용하고 있어서 일단 전자로 이해하고 default로 변경하겠습니다.

wheejuni replied:
아, 맨 밑의 `listPawnsToSingleLine` 만 말씀드린 겁니다. 이 메소드는 클래스 내부에서만 호출되는 것 같아서요.

wheejuni reviewed (changes requested):
안녕하세요, 어느덧 3단계네요. 구현하시느라 고생 많으셨습니다. 💯 👍 
코드를 읽고 궁금한 점과, 개선할 점들이 발견되어 아래와 같이 리뷰 의견 남겼습니다. 확인해주세요.

kyupid committed:
Fix a test method that prints

- 초기화하는 메소드를 빠뜨려서 원하는 결과값이 나오지 않았었습니다.

kyupid committed:
Fix access modifier

- 접근제어자를 수정했습니다.

kyupid committed:
Simplify for loop

- forEach 문을 사용해서 for loop을 간략하게 변경했습니다.

kyupid committed:
Remove duplicated List<Pawn> board

- 색깔별로 ArrayList를 추가해주면서 기존에 있던 중복되는 List를 삭제했습니다.

kyupid committed:
Add a method to distinguish Pawns' color

- 기존에 있던 List가 삭제가 되면서 기존에 있던 add(), size(), findPawn()의 코드가 새로운 ArrayList인 whitePawns 와 blackPawns 에 따라 변경이 필요해졌습니다.

- 그에 따라서, 먼저 어떤 색상의 ArrayList인지 확인해주는 메소드가 없으면 코드의 중복이 불가피하므로 어떤 ArrayList인지 확인해주는 메소드를 따로 만듭니다.

- if 조건문에는 color 로 어떤 색상의 ArrayList인지 확인하도록 했습니다.

kyupid committed:
Fix add(), size() and findPawn to re-function normally from removing the variable `List<Pawn> board`

- 기존에 있던 List가 삭제가 되면서 기존에 있던 add(), size(), findPawn()의 코드가 새로운 ArrayList인 whitePawns 와 blackPawns 에 따라 변경이 필요해졌습니다.

- 그에 따라서, 먼저 어떤 색상의 ArrayList인지 확인해주는 메소드가 없으면 코드의 중복이 불가피하므로 어떤 ArrayList인지 확인해주는 메소드를 따로 만듭니다.

- if 조건문에는 color 로 어떤 색상의 ArrayList인지 확인하도록 했습니다.

kyupid committed:
Verify a fixed method `Board.java`

- 테스트 코드도 새로워진 프로덕트 코드 확인을 위해 재수정을 했습니다.

kyupid committed:
Move listPawnsToSingleLine() to lower place to improve readability

- private 접근 제어자이기 때문에 덜 신경쓰기 위해 연관된 메소드보다 아래로 위치시켰습니다.

wheejuni commented on src/main/java/net/honux/chess/Board.java:
```diff
@@ -3,46 +3,40 @@
     }
 
-    public void initialize() {
+    void initialize() {
```
코멘트 참고해주세요. ㅠㅠ `listPawnsToSingleLine()` 에 대한 코멘트였는데 제가 코멘트 위치를 잘못 선정한 것 같네요.
이 부분은 `public` 으로 그대로 가는 게 맞습니다. 아래 메소드는 `private` 으로 잘 바꿔 주셨네요. 💯

wheejuni commented on src/main/java/net/honux/chess/Board.java:
```diff
@@ -55,6 +49,15 @@ public String getBlankLine() {
     }
 
+    private String listPawnsToSingleLine(ArrayList<Pawn> whichGroupOfPawns) {
+        StringBuilder sb = new StringBuilder();
+
+        for (Pawn p : whichGroupOfPawns)
+            sb.append(p.getRepresentation());
+
+        return sb.toString();
+    }
```
음 뭔가 훨씬 깔끔해졌네요? 좋습니다! 👍

wheejuni reviewed (approved):
수정사항까지 반영하시느라 수고 많으셨습니다.
중간에 멘붕의 위기가 있었던 것 같은데 잘 넘어가셨네요. 조금 서툴러도 어느덧 4단계로 갈 준비가 된 것 같습니다. 💯 
4단계에서도 좋은 모습 기대합니다. 다시 한 번 수고 많으셨고, 이 PR은 merge하도록 할게요. 👍
